tutorials/detailed_model_reference.ipynb

```diff
@@ -0,0 +1,663 @@
+    "    pos_target: torch.Tensor,\n",
+    "    data: Data,\n",
+    "    scale_id: int,\n",
+    "    reference_multivector: torch.Tensor\n",
+    ") -> torch.Tensor:\n",
+    "\n",
+    "    pos_diff = pos_source[data[f'scale{scale_id}_interp_source']] - pos_target[data[f'scale{scale_id}_interp_target']]\n",
+    "    squared_pos_dist = torch.clamp(torch.sum(pos_diff ** 2, dim=-1), min=1e-16).view(-1, 1, 1)\n",
+    "\n",
+    "    denominator = scatter(1. / squared_pos_dist, data[f'scale{scale_id}_interp_target'].long(), dim=0, reduce='sum')\n",
+    "\n",
+    "    multivectors = scatter(\n",
+    "        multivectors[data[f'scale{scale_id}_interp_source']] / squared_pos_dist,\n",
+    "        data[f'scale{scale_id}_interp_target'].long(),\n",
+    "        dim=0,\n",
+    "        reduce='sum'\n",
+    "    ) / denominator\n",
+    "\n",
+    "    scalars = scatter(\n",
+    "        scalars[data[f'scale{scale_id}_interp_source']] / squared_pos_dist.view(-1, 1),\n",
+    "        data[f'scale{scale_id}_interp_target'].long(),\n",
+    "        dim=0,\n",
+    "        reduce='sum'\n",
+    "    ) / denominator.view(-1, 1)\n",
+    "\n",
+    "    multivectors = torch.cat((multivectors, multivectors_skip), dim=-2)\n",
+    "    scalars = torch.cat((scalars, scalars_skip), dim=-1)\n",
+    "\n",
+    "    return mlp(multivectors, scalars, reference_mv=reference_multivector)"
+   ],
+   "id": "692225270547a492",
+   "outputs": [],
+   "execution_count": 6
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "```TODO```",
+   "id": "86c7c79f3c53ead6"
+  },
+  {
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-07-12T12:16:45.007227Z",
+     "start_time": "2024-07-12T12:16:45.000335Z"
+    }
+   },
+   "cell_type": "code",
+   "source": [
+    "class PointCloudPooling(PointCloudPooling):\n",
+    "\n",
+    "    def message(\n",
+    "        self,\n",
+    "        x_j: torch.Tensor,\n",
+    "        pos_i: torch.Tensor,\n",
+    "        pos_j: torch.Tensor,\n",
+    "        scalars_j: torch.Tensor,\n",
+    "        reference_multivector_j: torch.Tensor\n",
+    "    ) -> torch.Tensor:\n",
+    "\n",
+    "        multivectors, scalars = self.mlp(\n",
+    "            torch.cat((x_j, embed_translation(pos_j - pos_i).unsqueeze(-2)), dim=-2),\n",
+    "            scalars=scalars_j,\n",
+    "            reference_mv=reference_multivector_j\n",
+    "        )\n",
+    "\n",
+    "        return multivectors, scalars\n",
+    "\n",
+    "    def aggregate(self, inputs: tuple, index: torch.Tensor, ptr=None, dim_size=None) -> torch.Tensor:\n",
+    "        multivectors, scalars = (self.aggr_module(tensor, index, ptr=ptr, dim_size=dim_size, dim=self.node_dim) for tensor in inputs)\n",
+    "\n",
+    "        return multivectors, scalars"
+   ],
+   "id": "dc8b7295bcc0eebb",
+   "outputs": [],
+   "execution_count": 7
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "```TODO```",
```

Comment:
The complete tokenization module now just wraps pooling and interpolation. For reflection equivariance, we have to keep track of a reference multivector. Furthermore, we support extracting a global class token for classification tasks. This is a lot more efficient than taking the global average of vertex-wise outputs.